src/extensions/templateExtension.ts

```diff
@@ -0,0 +1,91 @@
+import { TemplateContext, ScopeData } from "../compilation";
+import { ITemplateExtension } from "./ITemplateExtension";
+import { IExtensionUtilities } from "./IExtensionUtilities";
+import { XmlDepthTracker, XmlNode } from "../xml";
+import { toDictionary } from "../utils";
+import { ITemplatePlugin } from "./ITemplatePlugin";
+import { IMap } from "../../src/types";
+
+/* eslint-disable @typescript-eslint/member-ordering */
+
+export abstract class TemplateExtension implements ITemplateExtension {
+    protected utilities: IExtensionUtilities;
+    private readonly plugins: ITemplatePlugin[];
+    protected readonly pluginsLookup: IMap<ITemplatePlugin>;
+
+    public maxXmlDepth = 20;
+
+    public abstract isMatch(node: XmlNode): boolean;
+    public abstract updateNode(node: XmlNode, data: ScopeData): void;
+    public abstract getXmlDocuments(
+        templateContext: TemplateContext
+    ): Promise<Map<string, XmlNode>>;
+
+    constructor(plugins: ITemplatePlugin[]) {
+        this.plugins = plugins;
+        this.pluginsLookup = toDictionary(plugins, p => p.contentType);
+    }
+
+    public async execute(
+        utilities: IExtensionUtilities,
+        data: ScopeData,
+        context: TemplateContext
+    ) {
+        this.utilities = utilities;
+        this.setUtilities(this.utilities);
+
+        const documents = await this.getXmlDocuments(context);
+
+        documents.forEach(document => {
+            const nodes = this.findNodes(document);
+
+            nodes.forEach(node => this.updateNode(node, data));
+        });
+    }
+
+    public setUtilities(utilities: IExtensionUtilities): void {
```

Comment:
Please elevate `setUtilities` to the Extension level. What I mean by that is the the TemplateCompiler should invoke this method, same way it's done with existing TemplatePlugins.

Reply:
Agreed